### PR TITLE
Add special cases to SA1129 code fix

### LINKED ISSUE #1972: Improve SA1129 code fix for enum types
Currently the SA1129 code fix replaces `new TaskState()` with `default(TaskState)`. An improved code fix for enum types would be the following:

* If a single member `Member` in the enum declaration `T` has the value 0, replace `new T()` with `T.Member`
* Otherwise, replace `new T()` with `default(T)` (i.e. fall back to the current behavior)

### PR DESCRIPTION
**Pull request by nbarbettini**
Fixes #1972 and #1973

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs [hunk 1]
```diff
@@ -58,12 +58,32 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+            }
+
+            return replacement
                 .WithLeadingTrivia(newExpression.GetLeadingTrivia())
                 .WithTrailingTrivia(newExpression.GetTrailingTrivia());
         }
 
+        private static SyntaxNode GetCancellationTokenNoneSyntax()
+        {
+            return SyntaxFactory.MemberAccessExpression(
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: This needs to start as fully-qualified and then be simplified (using `Simplifier`) rather than assuming `CancellationToken` is in scope.

Comment (nbarbettini) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: Ok, that makes sense.

**Comment by nbarbettini** [comment]
@sharwell I used your second suggestion and reused the `NameSyntax` from the constructor. How does this look?

**Comment by sharwell** [comment]
This seems to take care of it. I'll wait for @vweijsters or @pdelvo to second that before merging. :smile:

**Comment by Noryoko** [comment]
Should this support aliased cancellation tokens?

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs [hunk 2]
```diff
@@ -58,12 +58,32 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
         private static SyntaxNode GetReplacementNode(SyntaxNode node)
         {
             var newExpression = (ObjectCreationExpressionSyntax)node;
+            var nameSyntax = newExpression.Type as NameSyntax;
+            var identifierName = NameSyntaxHelpers.ToNormalizedString(nameSyntax);
 
-            return SyntaxFactory.DefaultExpression(newExpression.Type)
+            SyntaxNode replacement = null;
+            if (identifierName.EndsWith(nameof(CancellationToken), System.StringComparison.OrdinalIgnoreCase))
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: :exclamation: I overlooked this part originally. This check needs to check the *symbol*, not the syntax.

Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: The `EndsWith` check is also a problem when there is something like `new MyOwnCancellationToken()`. It should be a verify specific check on `CancellationToken`.

**Comment by nbarbettini** [comment]
@sharwell @vweijsters My latest push should address the issues you pointed out: the code fix now compares the symbol, not the syntax, and takes the fully qualified name (including namespace) into account. I've added a new unit test to verify that somebody adding a goofy `struct CancellationToken` won't trick it. :smile:

**Comment by nbarbettini** [comment]
@Noryoko Not sure what you mean - can you give an example?

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs [hunk 3]
```diff
@@ -81,10 +131,11 @@ private class FixAll : DocumentBasedFixAllProvider
                 }
 
                 var syntaxRoot = await document.GetSyntaxRootAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
+                var semanticModel = await document.GetSemanticModelAsync().ConfigureAwait(false);
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: This should use `fixAllContext.CancellationToken`.

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs [hunk 4]
```diff
@@ -81,10 +131,11 @@ private class FixAll : DocumentBasedFixAllProvider
...
 
                 var nodes = diagnostics.Select(diagnostic => syntaxRoot.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true));
 
-                return syntaxRoot.ReplaceNodes(nodes, (originalNode, rewrittenNode) => GetReplacementNode(rewrittenNode));
+                return syntaxRoot.ReplaceNodes(nodes, (originalNode, rewrittenNode) => GetReplacementNode(rewrittenNode, semanticModel, CancellationToken.None));
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: This should use `fixAllContext.CancellationToken`

**Comment by Noryoko** [comment]
@nbarbettini Here's an example.
```
using SystemToken = System.Threading.CancellationToken;

public class TestClass
{
    private SystemToken cancellantionToken = new SystemToken();
}
```

**Comment by nbarbettini** [comment]
@Noryoko I'll leave that up to @sharwell and @vweijsters. I'm not sure if it's worth supporting that edge case.

**Comment by sharwell** [comment]
@nbarbettini Now that you are using a semantic check, that case should actually be supported with no additional work. Maybe just add a unit test to verify.

**Comment by nbarbettini** [comment]
@sharwell Hey, you're right! :smile: Done.

**Comment by nbarbettini** [comment]
Since they were similar, I expanded this PR to include the code fix for #1972 as well. :+1:

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs [hunk 5]
```diff
@@ -49,21 +50,116 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                .OfType<IFieldSymbol>()
+                .Where(fs => fs.ConstantValue.Equals(0))
+                .ToList();
+
+            if (foundMembers.Count != 1)
+            {
+                return false;
+            }
+
+            foundMemberName = foundMembers.Single().Name;
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: :thought_balloon:  Since `foundMembers` is already a list, `foundMembers[0].Name` will work more efficiently.

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs [hunk 6]
```diff
@@ -49,21 +50,116 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+        /// Gets a qualified member access expression for the given <paramref name="typeSyntax"/>.
+        /// </summary>
+        /// <param name="typeSyntax">The type syntax from the original constructor.</param>
+        /// <param name="memberName">The member name.</param>
+        /// <returns>A new member access expression.</returns>
+        private static SyntaxNode ConstructMemberAccessSyntax(TypeSyntax typeSyntax, string memberName)
+            => SyntaxFactory.MemberAccessExpression(
+                SyntaxKind.SimpleMemberAccessExpression,
+                typeSyntax,
+                SyntaxFactory.IdentifierName(memberName));
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: :question: I'm not particularly found of this syntax.
@sharwell @pdelvo How do you think above the construction above?

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1129UnitTests.cs [hunk 7]
```diff
@@ -237,6 +237,292 @@ public T TestMethod2<T>()
...
+            DiagnosticResult[] expected =
+            {
+                this.CSharpDiagnostic().WithLocation(5, 18),
+            };
+
+            await this.VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None).ConfigureAwait(false);
+            await this.VerifyCSharpDiagnosticAsync(fixedTestCode, EmptyDiagnosticResults, CancellationToken.None).ConfigureAwait(false);
+            await this.VerifyCSharpFixAsync(testCode, fixedTestCode).ConfigureAwait(false);
+        }
+
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1129UnitTests.cs]: :question: Could you please also add a test case for an `enum` with multiple `0` value members?

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs (hunk 6 above)
Comment (nbarbettini) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: @vweijsters The expression-bodied method syntax?

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs (hunk 5 above)
Comment (nbarbettini) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: Good catch, thanks!

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1129UnitTests.cs (hunk 7 above)
Comment (nbarbettini) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1129UnitTests.cs]: Done

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs (hunk 6 above)
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: Yes, I think it reduces readability, but that is my personal opinion. That's why I asked @sharwell and @pdelvo for their opinions.

Comment (nbarbettini) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: I was on the fence in this case. Usually I think it improves readability, but with the large block here I wasn't sure.

Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1129CodeFixProvider.cs]: I personally wouldn't use the arrow syntax if the expression spans multiple lines. But I don't think it is a big deal.

**Comment by vweijsters** [comment]
:+1: Looks good to me now
